pkg/engine/decisions.go

```diff
@@ -0,0 +1,226 @@
+			if err != nil {
+				context.Errors = append(context.Errors, &InternalError{
+					Cause: err,
+					Child: &ResourceConfigurationError{
+						Resource: decision.Result.Resource,
+						Config:   *decision.Result.Config,
+					},
+				})
+				return
+			}
+			context.recordDecision(decision)
+		}
+	case ActionDelete:
+		if decision.Result.Resource != nil {
+			if context.ResourceGraph.RemoveResource(decision.Result.Resource) == nil {
+				err := context.ResourceGraph.RemoveResource(decision.Result.Resource)
+				if err != nil {
+					context.Errors = append(context.Errors, &InternalError{
+						Cause: err,
+					})
+					return
+				}
+				context.recordDecision(decision)
+			}
+		}
+	case ActionCreate:
+		if decision.Result.Resource != nil {
+			if context.ResourceGraph.GetResource(decision.Result.Resource.Id()) == nil {
+				context.ResourceGraph.AddResource(decision.Result.Resource)
+				context.recordDecision(decision)
+			}
+		}
+	case ActionConnect:
+		if decision.Result.Edge != nil {
+			if context.ResourceGraph.GetResource(decision.Result.Edge.Source.Id()) == nil {
+				e.handleDecision(context, Decision{Level: LevelInfo, Action: ActionCreate, Result: &DecisionResult{Resource: decision.Result.Edge.Source}, Cause: decision.Cause})
+			}
+			if context.ResourceGraph.GetResource(decision.Result.Edge.Destination.Id()) == nil {
+				e.handleDecision(context, Decision{Level: LevelInfo, Action: ActionCreate, Result: &DecisionResult{Resource: decision.Result.Edge.Destination}, Cause: decision.Cause})
+			}
+			if context.ResourceGraph.GetDependency(decision.Result.Edge.Source.Id(), decision.Result.Edge.Destination.Id()) == nil {
+				context.ResourceGraph.AddDependencyWithData(decision.Result.Edge.Source, decision.Result.Edge.Destination, decision.Result.Edge.Properties.Data)
+				context.recordDecision(decision)
+			}
+		}
+	case ActionDisconnect:
+		if decision.Result.Edge != nil {
+			if context.ResourceGraph.GetDependency(decision.Result.Edge.Source.Id(), decision.Result.Edge.Destination.Id()) != nil {
+				err := context.ResourceGraph.RemoveDependency(decision.Result.Edge.Source.Id(), decision.Result.Edge.Destination.Id())
+				if err != nil {
+					context.Errors = append(context.Errors, &InternalError{
+						Cause: err,
+					})
+					return
+				}
+				context.recordDecision(decision)
+			}
+		}
+	default:
+		panic("unknown action")
+	}
+}
+
+// func (d Decision) MarshalJSON() ([]byte, error) {
+// 	fmt.Println(`{"level":"` + string(d.Level) + `","result":"` + d.Result.String() + `","action":"` + string(d.Action) + `","cause":"` + d.Cause.String() + `"}`)
+// 	return []byte(`{"level":"` + string(d.Level) + `","result":"` + d.Result.String() + `","action":"` + string(d.Action) + `","cause":"` + d.Cause.String() + `"}`), nil
+// }
+
+func (r DecisionResult) MarshalJSON() ([]byte, error) {
+	if r.Config != nil {
```

Comment:
Use an internal type for these, should make it much cleaner. Though if the actual type used `ResourceId`s then this may not even be necessary.